classes/Tools.php

```diff
@@ -4264,10 +4264,11 @@ public static function getPath($url_base, $id_category, $path = '', $highlight =
                 $n = 1;
                 $n_categories = (int) count($categories);
                 foreach ($categories as $category) {
-                    $link = Context::getContext()->link->getAdminLink('AdminCategories');
-                    $edit = '<a href="' . Tools::safeOutput($link . '&id_category=' . (int) $category['id_category'] . '&' . (($category['id_category'] == 1 || $home) ? 'viewcategory' : 'updatecategory')) . '" title="' . ($category['id_category'] == Category::getRootCategory()->id_category ? 'Home' : 'Modify') . '"><i class="icon-' . (($category['id_category'] == Category::getRootCategory()->id_category || $home) ? 'home' : 'pencil') . '"></i></a> ';
+                    $action = (($category['id_category'] == (int) Configuration::get('PS_HOME_CATEGORY') || $home) ? 'admin_categories_index' : 'admin_categories_edit');
```

Comment:
Nice work @PululuK Although in this case you can take advantage of the `_legacy_link` feature https://devdocs.prestashop.com/1.7/development/architecture/migration-guide/controller-routing/#the-legacy-link-property

Which is able to automatically convert a legacy link into a symfony route, provided you correctly use the arguments and not copy paste them by hand. So you could use the actions `index` and `updatecategory` in place of (respectively) the routes `admin_categories_index` and `admin_categories_edit`

This feature also manages an automatic redirection even with urls formatted like legacy, however in this case it doesn't work because the mapping of the `id_category` is missing for the `admin_categories_index` route (probably because the parameter was managed after the initial creation).

Anyway I believe you could fix this whole bug just fixing the routing configuration in `src/PrestaShopBundle/Resources/config/routing/admin/sell/catalog/categories.yml`

```yaml
admin_categories_index:
    path: /{categoryId}
    methods: [GET, POST]
    defaults:
        _controller: PrestaShopBundle:Admin\Sell\Catalog\Category:index
        _legacy_controller: AdminCategories
        _legacy_link: AdminCategories
        _legacy_parameters:
            id_category: categoryId
        categoryId: null
    requirements:
        categoryId: '\d+'
```
Only by adding the missing `_legacy_parameters` part

Reply:
Hi @jolelievre 
Thank for you feedback ....


Very interesting your proposal ... Actually it lacks the `_legacy_parameters` (in `src/PrestaShopBundle/Resources/config/routing/admin/sell/catalog/categories.yml`) directive, adding this will prevent the redirection from always being on the category home, but this will not solve this problem for editing (which is the case with this issue : xxx).

Just with that :

```yml
admin_categories_edit:
    path: /{categoryId}/edit
    methods: [GET, POST]
    defaults:
        _controller: PrestaShopBundle:Admin\Sell\Catalog\Category:edit
        _legacy_controller: AdminCategories
        _legacy_link: AdminCategories:updatecategory
        _legacy_parameters:
            id_category: categoryId
    requirements:
        categoryId: \d+
```

`index.php/sell/catalog/categories?_token=xxxx&id_category=xxx&updatecategory` should redirect here `/index.php/sell/catalog/categories/xxx/edit?_token=xxxxx`, But this is not the case ... see https://github.com/PrestaShop/PrestaShop/issues/17908

I think in this case the problem is much more complex then ... can you open this PR please? I was wrong

Reply:
Oh you're right!! That's because the initial call to `getAdminLink` doesn't have the action parameter So it converts it to the index route (which makes sense), and then append the additional parameter

This would work well if the parameters were inject completely in `getAdminLink` So I guess the solution is a mix between what you did and fixing the routing

Once the routing is fixed, then you can really use the parameter `action` instead of `route` so you would need to replace:

```php
$action = (($category['id_category'] == (int) Configuration::get('PS_HOME_CATEGORY') || $home) ? 'index' : 'updatecategory');
```

and 

```php
$link = Context::getContext()->link->getAdminLink('AdminCategories',true, array('action'=> $action,'categoryId'=> (int) $category['id_category']));
```

Reply:
It also works, I just tested! Could you tell me how it is faster than injecting symfony parameters directly? Because it seems more logical to directly inject the parameters for the new controller instead of legacy. See https://github.com/PrestaShop/PrestaShop/blob/1.7.6.x/classes/Link.php#L705

Reply:
Done !

Reply:
The sf route params were initially introduced to allow creating symfony url through the `Link::getAdminLink` function (at the time the `_legacy_link` feature didn't exist)

Whereas the `_legacy_link` feature along with the `LegacyUrlConverter` service was specifically designed from a migration point of view. It has two main advantages:
- once the `_legacy_link` is correctly set the legacy urls are automatically redirected (when they find a matching route), so that even hard coded urls are correctly redirected to the new migrated page
- for modules developer you can use the getAdminLink in any version, if the controller has been migrated it will output the symfony url, if not it will output the legacy controller one

The only drawback (as you could see), is when parameters are concatenated after the getAdminLink call, because they are concatenated after the symfony url. We have no solution for this problem today, except correctly using the function parameters (which is sadly not possible in some early 1.7 versions).


As for this special case, since it's inside the core you are right: it doesn't bring much values It might be slightly less performant because it needs to find the matching route, but some cache policy has been implemented so the matching process is not that long

You might wonder, what is the advantage because this small performance drawback then? If I could inject the appropriate route directly. I would see two side advantages:
- if the symfony route is modified somehow (its name, parameters, ...) the modification would be automatically updated via this call
- for people using the previous versions it also gives them a valid example of how they should use the getAdminLink correctly and how they could integrate it (wether in the core or in their module)

Also, but this one is personal, I don't think mixing symfony routes inside the legacy classes is a good thing to do if we can avoid it.

Reply:
Thank you @jolelievre  ! Done